### ISSUE TITLE
proc execution

### ISSUE THREAD

**Issue by willox** [wixoaGit/OpenDream#94]
Currently procs are just executed with code like the following:
```csharp
Task.Run(() => connection.ClientDreamObject?.CallProc("North"));
```

This is no good because
1) The called code isn't thread safe
2) The DM code we've compiled will never realistically be parallel-safe
3) We need better control over the order that procs execute in, when they sleep, when they resume

My proposed solution for this is to create a simple executor that'll manage all of this for us and call everything inside the main thread. It'll take a couple of changes:

**The interpreter**
Convert `DreamProc::Run` into a coroutine that executes over the bytecode and yields when the proc should pause/crash/return. The points where that'll happen are fairly obvious, you'd obviously yield on `return` and on runtimes, but also on calls to asynchronous built-ins (sleep, spawn, etc.) and also when a proc without `waitfor = 0` that is being called by another proc yields. (That last bit gets a little tricky, but I've worked with code that emulated the same behaviour in C++/asm before so I think I could figure it out.)

**The executor**
The executor will maintain a list (the details there don't matter, it could be another structure) containing the next procs to resume and when to resume them. This would only be used for procs that are sleeping as other paused procs can be resumed or queued once the thing they are yielding on returns. Whenever a proc yields and we find ourselves with nothing to do, we simply call into the next proc in that list _or_ move to the next tick of the program's main loop based on some logic I haven't thought out.

**Coroutines?**
I've only worked with coroutines in C# using `IEnumerable` and `yield`. Maybe something similar can be done with tasks? I think it'd be easy enough to use the former though because I've done it before.

Alternatively the code can be written to be resumable without any fancy coroutine/task syntax. This approach might be ideal as it'll make inspecting the state of the VM (for debugging) fairly simple.

I can code it all but I want @wixoaGit's input first.

**Comment by wixoaGit** [comment]
Right now, calling a proc and getting its return value looks like `DreamValue returned = dreamObject.CallProc("proc")`. Would this have to be changed? What would happen if the called proc `sleep`s?

> Alternatively the code can be written to be resumable without any fancy coroutine/task syntax. This approach might be ideal as it'll make inspecting the state of the VM (for debugging) fairly simple.

I'm leaning towards doing this over using `yield` or Tasks.

**Comment by willox** [comment]
> I'm leaning towards doing this over using yield or Tasks.
That's good, because I pretty much had the same thoughts as I finished writing the issue (that was the last line I wrote.)

> Right now, calling a proc and getting its return value looks like DreamValue returned = dreamObject.CallProc("proc"). Would this have to be changed? What would happen if the called proc sleeps?

I'll try to explain what I think my approach to this would be, but I wouldn't know for sure until I code it.

First we need to define where we'll store the state about currently executing code:

**Call Context**
Multiple contexts can exist at once. We make a new context whenever we:
- Call a proc from the root-level in C# (somewhere like Program.cs). This includes any situation where the proc is not returning to another proc.
- Call `spawn`. As the body of a spawn statement does not return to the caller, it belongs on a new call stack.
- Performing a yielding operation (`sleep`, `input`, proc calls, etc.),  in a proc with `waitfor = 0`. In this case, `.` is returned to the caller and the current state of the proc is moved into a new stack.

Contexts would include a stack of procs, so that when the top-level proc returns its return value can be passed to the next proc in the stack (which then continues executing.)

If we perform a yielding operation in a proc that does **not** have `waitfor = 0`, the context yields and could be resumed in different ways (it'll be up to the operation that yielded the context to later resume it.)

In the case of `sleep`, the flow would probably be something like
1. Sleep opcode checks if the current proc has `waitfor = 0`. If it does, it performs a `detach` operation that
    1. Creates a new context containing only the current proc and its state.
    2. Sets the state of the current context as if our proc had just performed `return .`
2. Sleep opcode adds the context (or new context) into some sort of list that contains all currently sleeping procs. This is specific to sleeping (or spawned) procs, not things that will resume based on user-input or network transfers as they'll live somewhere else.
3. Now the coroutine-interpreter for our proc has to yield. It can yield for two reasons (`Returned` or `Yielded`), so we'll pass through an enum specifying the reason.
    1. `Returned` means that the next proc at the top of the context's stack should immediately continue execution (with our returned value it has been waiting for.)
    2. `Yielded` means that the context should stop executing. It'll resume later when something wants it to (in this case, it'll be some logic somewhere that checks our list of sleeping procs.)
4. That enum, through whatever the code path ends up being, ends up in a function that I'll call `CallContext::Execute`. There'll be two paths it can take from here
    1. If `Returned`, this is where we start executing that next proc in the stack I mentioned
    2. If `Yielded`, `CallContext::Execute` returns and everything is done.
5. Once `CallContext::Execute` eventually returns, the executor/scheduler (the thing that would have called `CallContext::Execute`) starts running a new proc (potentially from the list of sleeping prosc), or if we've run out of time for this tick will advance to the next tick.

**Comment by willox** [comment]
https://github.com/wixoaGit/OpenDream/compare/master...willox:execution_context


I've made a little progress on this tonight, but there's more to do so I'll have to finish another time.

In cases where you don't care about asynchronous results, calling a proc looks like this (although of course it can and will be in a wrapper):
```csharp
var proc = FindProc(...);

var context = new ExecutionContext();
context.PushNewProcState(proc, instance, usr, arguments);

DreamValue? value = context.Resume(); // Null indicates that the proc yielded instead of returning, we don't have to do anything for the proc to continue its asynchronous execution.
```

I haven't implemented native procs yet, so that's the main thing I'll have to do tomorrow. They'll be able to have state and yield if they want instead of being synchronous.

If you want to see how I've handled the state for all of this, just check out DreamProc.cs.
